pkg/reconciler/inmemorychannel/inmemorychannel.go

```diff
@@ -0,0 +1,338 @@
+		}
+		return err
+	}
+	imc.Status.PropagateDispatcherStatus(&d.Status)
+
+	// Get the Dispatcher Service and propagate the status to the Channel in case it does not exist.
+	// We don't do anything with the service because it's status contains nothing useful, so just do
+	// an existence check. Then below we check the endpoints targeting it.
+	_, err = r.serviceLister.Services(r.dispatcherNamespace).Get(r.dispatcherServiceName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			imc.Status.MarkServiceFailed("DispatcherServiceDoesNotExist", "Dispatcher Service does not exist")
+		} else {
+			logging.FromContext(ctx).Error("Unable to get the dispatcher service", zap.Error(err))
+			imc.Status.MarkServiceFailed("DispatcherServiceGetFailed", "Failed to get dispatcher service")
+		}
+		return err
+	}
+
+	imc.Status.MarkServiceTrue()
+
+	// Get the Dispatcher Service Endpoints and propagate the status to the Channel
+	// endpoints has the same name as the service, so not a bug.
+	e, err := r.endpointsLister.Endpoints(r.dispatcherNamespace).Get(r.dispatcherServiceName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			imc.Status.MarkEndpointsFailed("DispatcherEndpointsDoesNotExist", "Dispatcher Endpoints does not exist")
+		} else {
+			logging.FromContext(ctx).Error("Unable to get the dispatcher endpoints", zap.Error(err))
+			imc.Status.MarkEndpointsFailed("DispatcherEndpointsGetFailed", "Failed to get dispatcher endpoints")
+		}
+		return err
+	}
+
+	if len(e.Subsets) == 0 {
+		logging.FromContext(ctx).Error("No endpoints found for Dispatcher service", zap.Error(err))
+		imc.Status.MarkEndpointsFailed("DispatcherEndpointsNotReady", "There are no endpoints ready for Dispatcher service")
+		return errors.New("there are no endpoints ready for Dispatcher service")
+	}
+
+	imc.Status.MarkEndpointsTrue()
+
+	// Reconcile the k8s service representing the actual Channel. It points to the Dispatcher service via
+	// ExternalName
+	svc, err := r.reconcileChannelService(ctx, imc)
+	if err != nil {
+		imc.Status.MarkChannelServiceFailed("ChannelServiceFailed", fmt.Sprintf("Channel Service failed: %s", err))
+		return err
+	}
+	imc.Status.MarkChannelServiceTrue()
+	imc.Status.SetAddress(fmt.Sprintf("%s.%s.svc.%s", svc.Name, svc.Namespace, utils.GetClusterDomainName()))
+
+	// Ok, so now the Dispatcher Deployment & Service have been created, we're golden since the
+	// dispatcher watches the Channel and where it needs to dispatch events to.
+	return nil
+}
+
+func (r *Reconciler) reconcileChannelService(ctx context.Context, imc *v1alpha1.InMemoryChannel) (*corev1.Service, error) {
+	// Get the  Service and propagate the status to the Channel in case it does not exist.
+	// We don't do anything with the service because it's status contains nothing useful, so just do
+	// an existence check. Then below we check the endpoints targeting it.
+	svc, err := r.serviceLister.Services(imc.Namespace).Get(fmt.Sprintf("%s-kn-channel", imc.Name))
```

Comment:
This may be too long, possibly truncate.

Also, is it intended that this is now the standard channel service name? If not, add a comment that this can change and people still need to follow the Addressable contract.